go/package/download.php

```diff
@@ -0,0 +1,152 @@
+<?php
+  declare(strict_types=1);
+
+  namespace Keyman\Site\com\keyman;
+
+  require __DIR__ . '/../../_includes/autoload.php';
+
+  const DEBUG=0;
+
+  use Keyman\Site\com\keyman\Validation;
+  use Keyman\Site\com\keyman\KeymanHosts;
+
+  PackageDownloadPage::redirect_to_file(
+    isset($_REQUEST['id']) ? $_REQUEST['id'] : null,
+    isset($_REQUEST['version']) ? $_REQUEST['version'] : null,
+    isset($_REQUEST['platform']) ? $_REQUEST['platform'] : null,
+    isset($_REQUEST['tier']) ? $_REQUEST['tier'] : null,
+    isset($_REQUEST['bcp47']) ? $_REQUEST['bcp47'] : null,
+    isset($_REQUEST['update']) ? $_REQUEST['update'] : null,
+    isset($_COOKIE['_ga']) ? $_COOKIE['_ga'] : null
+  );
+
+  class PackageDownloadPage {
+
+    public static function redirect_to_file($id, $version, $platform, $tier, $bcp47, $update, $ga_cookie) {
+      if(empty($id)) {
+        echo "Invalid parameters; id and version expected";
+        exit;
+      }
+
+      if(empty($version)) {
+        // If version isn't provided, we'll query the live api for the version
+        $json = @file_get_contents(KeymanHosts::Instance()->api_keyman_com . '/keyboard/' . rawurlencode($id));
+        if ($json !== FALSE) {
+          $json = json_decode($json);
+        }
+
+        if(empty($json)) {
+          echo "Invalid parameter; id not found";
```

Comment:
Shouldn't this be more like the following?

```suggestion
          echo "Version parameter missing and could not be found automatically";
```

This `$json` corresponds to a version lookup, so why emit an error saying the _id_ couldn't be found here?

Reply:
If the api returns a fail, then the most likely reason is that the id we passed in was invalid. IMO other error conditions are so rare as to not worry about. 

It just so happens that we are only validating id when we do a version lookup, because invalid id will come out in the wash after the redirect anyway otherwise (but we need an exit route in this situation).

So the error message is reporting back to the page user a message that helps them understand that the parameters they passed were invalid. Perhaps it could be better worded, but the issue to the user is not around the version parameter, it's around the id parameter.